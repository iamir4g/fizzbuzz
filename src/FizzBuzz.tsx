import { useEffect, useRef, useState } from "react";

export const InputNumber = ({
  buttonText,
  onSubmit,
}: {
  buttonText: string;
  onSubmit: (x: number) => void;
}) => {
  const ref = useRef<HTMLInputElement>(null);
  useEffect(() => {
    ref.current?.focus();
  }, []);
  const [value, setValue] = useState(0);
  return (
    <div>
      <div>
        <input
          type="number"
          min={0}
          ref={ref}
          value={value}
          onChange={(event) => setValue(parseInt(event.target.value))}
        ></input>
        <button onClick={() => onSubmit(value)}>{buttonText}</button>
      </div>
    </div>
  );
};

type FizzBuzzState =
  | { _tag: "Input" }
  | { _tag: "started"; startNumber: number };

export const FizzBuzzInternal = ({ startNumber }: { startNumber: number }) => {
  const [value, setValue] = useState(startNumber);
  const divisiableBy3 = value % 3 === 0;
  const divisiableBy5 = value % 5 === 0;
  return (
    <div>
      <button onClick={() => setValue((old) => old + 1)}>+</button>
      <b data-testid="fizzbuzz_text">
        {divisiableBy3 && divisiableBy5
          ? "Fizz Buzz"
          : divisiableBy3
          ? "Fizz"
          : divisiableBy5
          ? "Buzz"
          : value}{" "}
      </b>
      <button onClick={() => setValue((old) => old - 1)}>-</button>
    </div>
  );
};
export const FizzBuzz = () => {
  const [state, setState] = useState<FizzBuzzState>({ _tag: "Input" });
  return (
    <div data-testid="fizzbuzz_container">
      {state._tag === "Input" ? (
        <InputNumber
          onSubmit={(x) => setState({ _tag: "started", startNumber: x })}
          buttonText="start"
        ></InputNumber>
      ) : (
        <FizzBuzzInternal startNumber={state.startNumber}></FizzBuzzInternal>
      )}
    </div>
  );
};
