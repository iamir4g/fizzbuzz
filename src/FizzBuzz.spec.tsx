import { act, fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import { FizzBuzz, FizzBuzzInternal, InputNumber } from "./FizzBuzz";
import exp from "constants";
describe("FizzBuzz", () => {
  describe("Input", () => {
    it("shoul show a button a number input and have focused ", () => {
      render(
        <InputNumber onSubmit={() => {}} buttonText="submit"></InputNumber>
      );
      const textBox = screen.getByRole("spinbutton");

      expect(textBox).toBeVisible();
      expect(textBox).toHaveAttribute("min", "0");
      expect(textBox).toHaveFocus();

      const button = screen.getByRole("button", { name: /submit/i });
      expect(button).toBeVisible();
    });

    it("when button click it shoud return a number", () => {
      const submitFn = jest.fn();
      render(
        <InputNumber onSubmit={submitFn} buttonText="submit"></InputNumber>
      );
      const textBox = screen.getByRole("spinbutton");
      act(() => {
        userEvent.type(textBox, "6");
      });

      const button = screen.getByRole("button", { name: /submit/i });
      fireEvent.click(button);

      expect(submitFn).toHaveBeenCalled();
      expect(submitFn).toHaveBeenCalledWith(6);
    });
  });

  describe("start fizzbuzz", () => {
    it("should show 1 after start", () => {
      render(<FizzBuzz></FizzBuzz>);
      const startButton = screen.getByRole("button", { name: /start/i });
      const textInput = screen.getByRole("spinbutton");

      act(() => {
        userEvent.type(textInput, "1");
      });

      fireEvent.click(startButton);
      const startButton_ = screen.queryAllByRole("button", { name: /start/i });
      expect(startButton_).toBeNull;

      const text = screen.getByText("1");
    });
  });

  describe("FizzBuzz Internall", () => {
    const clickPlus = (plus: HTMLElement, repeat: number) => {
      for (let i = 0; i < repeat; i++) {
        fireEvent.click(plus);
      }
    };
    const renderAndreturnInput = (startNumber: number) => {
      render(<FizzBuzzInternal startNumber={startNumber}></FizzBuzzInternal>);

      const plus = screen.getByRole("button", { name: "+" });
      const mins = screen.getByRole("button", { name: "-" });
      const text = screen.getByTestId("fizzbuzz_text");

      return { plus, mins, text };
    };
    it("shoud show two buttons", () => {
      const { mins, plus, text } = renderAndreturnInput(1);
      expect(plus).toBeVisible();
      expect(mins).toBeVisible();
      expect(text).toHaveTextContent("1");
    });
    it("should incremnet and decrement with plus and minus", () => {
      const { mins, plus, text } = renderAndreturnInput(1);
      fireEvent.click(plus);
      expect(text).toHaveTextContent("2");
      fireEvent.click(mins);
      expect(text).toHaveTextContent("1");
    });
    it("should show for 3 and 9", () => {
      const { mins, plus, text } = renderAndreturnInput(1);

      clickPlus(plus, 2);
      expect(text).toHaveTextContent("Fizz");
      clickPlus(plus, 6);
      expect(text).toHaveTextContent("Fizz");
    });
    it("should show for 5 and 10", () => {
      const { mins, plus, text } = renderAndreturnInput(1);

      clickPlus(plus, 4);
      expect(text).toHaveTextContent("Buzz");
      clickPlus(plus, 5);
      expect(text).toHaveTextContent("Buzz");
    });
    it("should show Fizz Buzz for 15", () => {
      const { mins, plus, text } = renderAndreturnInput(1);

      clickPlus(plus, 14);
      expect(text).toHaveTextContent("Fizz Buzz");
    });
  });
  describe("Snapshot testing", () => {
    it("should be ok", () => {
      render(<FizzBuzz></FizzBuzz>);

      const div = screen.getByTestId("fizzbuzz_container");
      expect(div).toMatchSnapshot();
    });
  });
});
