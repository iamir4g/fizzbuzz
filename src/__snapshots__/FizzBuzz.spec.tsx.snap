// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FizzBuzz Snapshot testing should be ok 1`] = `
<div
  data-testid="fizzbuzz_container"
>
  <div>
    <div>
      <input
        min="0"
        type="number"
        value="0"
      />
      <button>
        start
      </button>
    </div>
  </div>
</div>
`;
